src/main/java/htsjdk/samtools/util/AsyncWriterPool.java

```diff
@@ -0,0 +1,124 @@
+package htsjdk.samtools.util;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class AsyncWriterPool<A> implements Closeable {
+    private final ExecutorService executor;
+    private final AtomicBoolean poolClosed;
+    private final ArrayList<PooledWriter<A>> writers;
+
+    public AsyncWriterPool(final int threads) {
+        assert threads >= 1; // TODO: throw error
+
+        this.poolClosed = new AtomicBoolean(false);
+        // TODO: make this structure threadsafe
+        this.writers = new ArrayList<>();
+
+        executor = Executors.newWorkStealingPool(threads);
+    }
+
+    @Override
+    public void close() throws IOException {
+        this.poolClosed.set(true);
+        for (PooledWriter<A> writer : this.writers) {
+            writer.close();
+        }
+        this.executor.shutdown();
+    }
+
+    public static class PooledWriter<B> implements WrappedWriter<B> {
+        private final WrappedWriter<B> writer;
+        private final BlockingQueue<B> queue;
+        private final AtomicBoolean writing;
+        private final AtomicBoolean isClosed;
+        public final AsyncWriterPool<B> outer;
+
+        private final int buffsize;
+        private Future<Integer> currentTask;
+
+
+        public PooledWriter(final AsyncWriterPool<B> outer, final WrappedWriter<B> writer, final BlockingQueue<B> queue, final int buffSize) {
+            assert buffSize >= 1; // TODO: Add error message
+            this.writer = writer;
+            this.queue = queue;
+            this.buffsize = buffSize;
+            this.outer = outer;
+            this.writing = new AtomicBoolean(false);
+            this.isClosed = new AtomicBoolean(false);
+            this.currentTask = null;
+
+            // Add self to writer
+            this.outer.writers.add(this);
```

Comment:
I would do this in the outer class directly, it feels odd here.

Reply:
In order for AsyncPoolWriter to not be type based on the writer, the writer has to be instantiated in the inner class. Or at least I can't figure out how to do it any other way.